# Video Insights

In [1]:
import pandas as pd
import mysql.connector

In [2]:
from collections import Counter

In [3]:
pd.set_option('display.max_columns', None)

In [5]:
# database connection
cnx = mysql.connector.connect(host = 'mysqlservernewjprod.mysql.database.azure.com', user = "phantom@mysqlservernewjprod", password = 'Zurich$1', db = 'fb', port = 3306)

In [6]:
cursor = cnx.cursor(buffered=True)

In [7]:
sql = 'SELECT * FROM fb.videoinsights where pageName = "NEWJPLUS";' #sql query for retreiving the data

df = pd.read_sql(sql, cnx) # retreive the data

df.to_csv('videoinsights.csv', index=False) # save the retreived data locally

cursor.close() 

True

In [8]:
data = pd.read_csv('videoinsights.csv') # read the loacally saved data
data.head(3)

id          videoId  universal_video_id  \
0  46013c80-51b3-11ea-9de5-48682eb0e778  115834923200548              7161.0   
1  460a7d72-51b3-11ea-9de5-48682eb0e778  116029313162712                 NaN   
2  bbae9f98-51b2-11ea-9de5-48682eb0e778  117718152938744              6789.0   

             pageId  pageName  \
0  1954072204870360  newjplus   
1  1954072204870360  newjplus   
2  1954072204870360  newjplus   

                                       permalink_url  \
0  https://www.facebook.com/NEWJplus/videos/11583...   
1  https://www.facebook.com/NEWJplus/videos/11602...   
2  https://www.facebook.com/NEWJplus/videos/11771...   

                                               title  \
0  Shoaib Akhtar Reveals The Reason Behind Pakist...   
1  Arvind Kejriwal Sings Hanuman Chalisa On Natio...   
2  Mahatma Gandhi Death Anniversary - The Actuall...   

                                         description         created_time  \
0  शोएब अख्तर ने बताया पाकिस्तान क्रिकेट का सच\n\...  2020-02-06 11:13:05   
1  देखिए केजरीवाल का हनुमान भक्त रूप, जब गाई उन्ह...  2020-02-04 09:45:18   
2  जानिए उस दिन की पूरी कहानी जब महात्मा गांधी को...  2020-01-29 13:48:38   

  created_time_zone   published_time_ist  is_crosspost_video published  \
0            +00:00  2020-02-06 16:43:05               False       NaN   
1            +00:00  2020-02-04 15:15:18                True       NaN   
2            +00:00  2020-01-29 19:18:38               False       NaN   

    length  total_video_followers  total_video_views  \
0  182.719                      0            1571981   
1   80.639                      0               7045   
2  180.202                      0             109916   

   total_video_views_unique  total_video_views_autoplayed  \
0                   1474663                       1491601   
1                      7017                          6703   
2                    109916                        103821   

   total_video_views_clicked_to_play  total_video_views_organic  \
0                              80380                     679180   
1                                342                       7045   
2                               6095                     109916   

   total_video_views_organic_unique  total_video_views_paid  \
0                            655042                  892801   
1                              7017                       0   
2                            109916                       0   

   total_video_views_paid_unique  total_video_views_sound_on  \
0                         848487                     1489376   
1                              0                        6168   
2                              0                       90673   

   total_video_play_count  total_video_consumption_rate  \
0                 2633106                             1   
1                   17106                             0   
2                  249968                             0   

   total_video_complete_views  total_video_complete_views_unique  \
0                       98568                              98568   
1                         452                                451   
2                       18023                              18003   

   total_video_complete_views_auto_played  \
0                                   87894   
1                                     392   
2                                   15985   

   total_video_complete_views_clicked_to_play  \
0                                       10674   
1                                          60   
2                                        2038   

   total_video_complete_views_organic  \
0                               57154   
1                                 452   
2                               18023   

   total_video_complete_views_organic_unique  total_video_complete_views_paid  \
0                                      56950                            41414   
1                                        451                   

In [9]:
data.columns

Index(['id', 'videoId', 'universal_video_id', 'pageId', 'pageName',
       'permalink_url', 'title', 'description', 'created_time',
       'created_time_zone', 'published_time_ist', 'is_crosspost_video',
       'published', 'length', 'total_video_followers', 'total_video_views',
       'total_video_views_unique', 'total_video_views_autoplayed',
       'total_video_views_clicked_to_play', 'total_video_views_organic',
       'total_video_views_organic_unique', 'total_video_views_paid',
       'total_video_views_paid_unique', 'total_video_views_sound_on',
       'total_video_play_count', 'total_video_consumption_rate',
       'total_video_complete_views', 'total_video_complete_views_unique',
       'total_video_complete_views_auto_played',
       'total_video_complete_views_clicked_to_play',
       'total_video_complete_views_organic',
       'total_video_complete_views_organic_unique',
       'total_video_complete_views_paid',
       'total_video_complete_views_paid_unique',
       'tota

In [10]:
data.shape

(5540, 80)

In [14]:
missing_universal_id = data[data.universal_video_id.isnull()]

In [15]:
missing_universal_id.to_csv('missing_universal_id_videoinsights_data.csv', index=False)

In [16]:
missing_universal_id.shape

(2377, 80)

### Task 1: check the total number of unique video Id's

In [11]:
# check unique video id's
# the number of unique video id's means number of videos because each video id represent one unique video
data['videoId'].nunique()

5659

`Total 5659 videos data retreived`

### Task 2: check the last date for which data is retreived

In [12]:
# check the data type of 'video published time'
data['published_time_ist'].dtype 

dtype('O')

In [13]:
# covert the data['published_time_ist'] to datetime format
data['published_time_ist'] = pd.to_datetime(data['published_time_ist'])

In [14]:
# now check the last date for which data is retreived
data['published_time_ist'].max()

Timestamp('2020-10-19 22:05:01')

`Last date for which data was populated 19th oct 2020 which is (today-3)`

In [16]:
data['published_time_ist'].min() # Initial date for which we have data in database

Timestamp('2019-02-01 09:00:00')

###### Any video published before  data['published_time_ist'].max() - 45 days should have the 45 days data

### Task 3: Check if the videos published before 45 days have 45 observations or not

In [15]:
from datetime import datetime, timedelta

In [16]:
# check the date, before which any video published should have 45 days record
data['published_time_ist'].max() - timedelta(days = 45)

Timestamp('2020-09-04 22:05:01')

In [17]:
# create a dataframe of the videos for which data should be populated 
df_must_45_days = data[data['published_time_ist'] < data['published_time_ist'].max() - timedelta(days = 45)]

In [18]:
df_must_45_days['published_time_ist'].max() # max date

Timestamp('2020-09-04 21:45:00')

In [19]:
Counter(df_must_45_days['videoId'])

Counter({121907679248283: 45,
         131251801583746: 45,
         159309575071366: 45,
         169926994296431: 45,
         171037124192609: 45,
         177365873532288: 45,
         180124883239041: 45,
         181491583127053: 46,
         185784969170417: 45,
         188988899165178: 29,
         195832038433525: 46,
         197138025018068: 45,
         204377750880700: 45,
         228401808304999: 45,
         230281634927035: 45,
         234976301256198: 45,
         240165207239097: 45,
         245043266485381: 45,
         246147033338652: 43,
         267802947484871: 45,
         274134603629494: 45,
         276159787034729: 29,
         283732025942811: 45,
         306119667422427: 45,
         308204357102313: 45,
         322281705593049: 45,
         329656738404054: 45,
         336853384169775: 45,
         338251397054002: 45,
         347532586061866: 45,
         352198196176133: 45,
         360215928340814: 45,
         361098111232348: 45,
         3

In [20]:
index_id_less_than_45 = [i for i in range(len(list(Counter(df_must_45_days['videoId']).values()))) if list(Counter(df_must_45_days['videoId']).values())[i]<45]

In [21]:
ids_with_less_than_45 = [list(Counter(df_must_45_days['videoId']).keys())[i] for i in index_id_less_than_45]

In [22]:
len(ids_with_less_than_45)

185

`There are total 185 videos for which data isn't retreived for 45 days`

In [23]:
# check the list of the Id's for which data is not populated correctly
ids_with_less_than_45

[188988899165178,
 246147033338652,
 276159787034729,
 726264157916141,
 729109411190690,
 740372620057665,
 747734206000187,
 793020047918937,
 954315911693792,
 1192429324434437,
 1626054707567665,
 2459086174368958,
 2459951354282440,
 2465292237081685,
 2469208170023425,
 2479520118992230,
 2480601745550734,
 2637219943222246,
 2637300913214149,
 2637339306543643,
 2637408176536756,
 2637479939862913,
 2637995443144696,
 2638068189804088,
 2638173149793592,
 2638185716459002,
 2638244729786434,
 2638247226452851,
 2638282469782660,
 2638353259775581,
 2638868676390706,
 2639054839705423,
 2639197796357794,
 2639210999689807,
 2639237133020527,
 2639282789682628,
 2639825946294979,
 2639951012949139,
 2639992572944983,
 2640038009607106,
 2640172206260353,
 2640181039592803,
 2640185306259043,
 2640881016189472,
 2640926982851542,
 2640961896181384,
 2641016746175899,
 2674479769496263,
 2674531149491125,
 2674541859490054,
 2674600132817560,
 2674611349483105,
 2674623152815258,
 2

In [27]:
# check one of these Id's for verification
data[data['videoId'] == 188988899165178]

id          videoId            pageId  \
406  05faffec-b2d4-11ea-b179-ecc92973c78a  188988899165178  1954072204870360   
407  05fc2956-b2d4-11ea-b179-ecc92973c78a  188988899165178  1954072204870360   
408  05fcb1e5-b2d4-11ea-b179-ecc92973c78a  188988899165178  1954072204870360   
409  d66861db-b498-11ea-b179-ecc92973c78a  188988899165178  1954072204870360   
410  d668ec82-b498-11ea-b179-ecc92973c78a  188988899165178  1954072204870360   
411  a3cf72af-b62a-11ea-b179-ecc92973c78a  188988899165178  1954072204870360   
412  a3cffa32-b62a-11ea-b179-ecc92973c78a  188988899165178  1954072204870360   
413  e9e1d952-b7bc-11ea-b179-ecc92973c78a  188988899165178  1954072204870360   
414  e9e257f6-b7bc-11ea-b179-ecc92973c78a  188988899165178  1954072204870360   
415  f4b2965b-b94e-11ea-b179-ecc92973c78a  188988899165178  1954072204870360   
416  f4b309df-b94e-11ea-b179-ecc92973c78a  188988899165178  1954072204870360   
417  a0e08738-baaf-11ea-b179-ecc92973c78a  188988899165178  1954072204870360   
418  a0e10c7b-baaf-11ea-b179-ecc92973c78a  188988899165178  1954072204870360   
419  b6a45c83-bc41-11ea-b179-ecc92973c78a  188988899165178  1954072204870360   
420  b6a4dd39-bc41-11ea-b179-ecc92973c78a  188988899165178  1954072204870360   
421  d7c095a2-bdd3-11ea-b179-ecc92973c78a  188988899165178  1954072204870360   
422  d7c11e59-bdd3-11ea-b179-ecc92973c78a  188988899165178  1954072204870360   
423  070e295a-bf66-11ea-b179-ecc92973c78a  188988899165178  1954072204870360   
424  070eb371-bf66-11ea-b179-ecc92973c78a  188988899165178  1954072204870360   
425  3a870160-c0f8-11ea-b179-ecc92973c78a  188988899165178  1954072204870360   
426  3a878406-c0f8-11ea-b179-ecc92973c78a  188988899165178  1954072204870360   
427  6f1c17fe-c28a-11ea-a721-c08f8b0c98d5  188988899165178  1954072204870360   
428  6f1c9b18-c28a-11ea-a721-c08f8b0c98d5  188988899165178  1954072204870360   
429  99667bf7-c41c-11ea-a721-c08f8b0c98d5  188988899165178  1954072204870360   
430  99670676-c41c-11ea-a721-c08f8b0c98d5  188988899165178  1954072204870360   
431  ca870363-c5ae-11ea-a721-c08f8b0c98d5  188988899165178  1954072204870360   
432  ca878bab-c5ae-11ea-a721-c08f8b0c98d5  188988899165178  1954072204870360   
433  f7a6c259-c740-11ea-a721-c08f8b0c98d5  188988899165178  1954072204870360   
434  f7a7440a-c740-11ea-a721-c08f8b0c98d5  188988899165178  1954072204870360   

     pageName                                      permalink_url  \
406  newjplus  https://www.facebook.com/NEWJplus/videos/18898...   
407  newjplus  https://www.facebook.com/NEWJplus/videos/18898...   
408  newjplus  https://www.facebook.com/NEWJplus/videos/18898...   
409  newjplus  https://www.facebook.com/NEWJplus/videos/18898...   
410  newjplus  https://www.facebook.com/NEWJplus/videos/18898...   
411  newjplus  https://www.facebook.com/NEWJplus/videos/18898...   
412  newjplus  https://www.facebook.com/NEWJplus/videos/18898...   
413  newjplus  https://www.facebook.com/NEWJplus/videos/18898...   
414  newjplus  https://www.facebook.com/NEWJplus/videos/18898...   
415  newjplus  https://www.facebook.com/NEWJplus/videos/18898...   
416  newjplus  https://www.facebook.com/NEWJplus/videos/18898...   
417  newjplus  https://www.facebook.com/NEWJplus/videos/18898...   
418  newjplus  https://www.facebook.com/NEWJplus/videos/18898...   
419  newjplus  https://www.facebook.com/NEWJplus/videos/18898...   
420  newjplus  https://www.facebook.com/NEWJplus/videos/18898...   
421  newjplus  https://www.facebook.com/NEWJplus/videos/18898...   
422  newjplus  https://www.facebook.com/NEWJplus/videos/18898...   
423  newjplus  https://www.facebook.com/NEWJplus/videos/18898...   
424  newjplus  https://www.facebook.com/NEWJplus/videos/18898...   
425  newjplus  https://www.facebook.com/NEWJplus/videos/18898...   
426  newjplus  https://www.facebook.com/NEWJplus/videos/18898...   
427  newjplus  https://www.facebook.com/NEWJplus/videos/18898...   
428  newjplus  https://www.facebook.com/NEWJplus/videos/18898...   
429  n

In [28]:
data[data['videoId'] == 188988899165178].shape

(29, 34)

`This video has been updated for 29 days only`

In [29]:
# last day of video got updated
data[data['videoId'] == 188988899165178]['post_video_views_end_time_ist'].max()

'2020-07-15 12:30:00'

In [30]:
# check the date till when video should be updated
data[data['videoId'] == 188988899165178]['published_time_ist'].min() + timedelta(days = 45)

Timestamp('2020-08-01 15:25:28')

`Video should have been updated till 2020-08-01 but its last updated date is 2020-07-15`

### Task 4: Check if there is any missing data

In [31]:
data.isnull().sum()

id                                           0
videoId                                      0
pageId                                       0
pageName                                     0
permalink_url                                0
message                                      0
post_created_time                            0
post_created_time_zone                       0
published_time_ist                           0
post_video_views                             0
post_video_views_end_time                    0
post_video_views_end_time_ist                0
post_video_view_time                         0
post_video_view_time_end_time                0
post_video_view_time_end_time_ist            0
post_video_view_time_organic                 0
post_video_view_time_organic_end_time        0
post_video_view_time_organic_end_time_ist    0
post_video_views_10s_paid                    0
post_video_views_10s_paid_end_time           0
post_video_views_10s_paid_end_time_ist       0
post_video_vi

`no data is missing`